### DISCUSSION TITLE
Replace `delay` with `scheduled_for`?

### DISCUSSION THREAD

**Discussion by kirillsalykin** [maxcountryman/underway#56] [Q&A]
https://github.com/maxcountryman/underway/blob/main/migrations/20240921151751_0.sql#L50

Currently the delayed tasks are implemented via `delay` field.

Wdyt about adjusting this slightly, so it rely on `scheduler_for`.

so instead of  `and created_at + delay <= now()`, it will check `scheduled_for <= now()`?

The only reason it feels a bit "cleaner" for schedule for a specific time (and makes query a little simpler).

it is very subjective, thus currious about your opinion on this?

(of course it can also be solved with a fn that makes `delay` out of schedule_for), but this way (`schedule_for) "feels" cleaner for me for both schedule and dequeue)

**Comment by maxcountryman** [comment]
One of the reasons for using intervals is to ensure that all the time zone conversions happen in Postgres. Notice how we always use Postgres to create timestamps and only use spans that are converted to intervals as application-level parameters.

I think it would also be a little misleading to say it's scheduled for a particular wall clock instant because in reality it's not (the scheduler is required for that). Instead it's actually a duration before which the task can't be dequeued and after which it may or may not be.

So I generally think interval delays are a slightly more appropriate interface.

**Reply by kirillsalykin** [reply]
I am not saying that it should be scheduled for specific time, I mean it can be picked after the specific time.

Also I am no fully understand why `scheduler` is needed.

My point is, that instead of saying `do it in 1 hour`, I would prefer to say `do it at 12`.

So it is pretty boils down to replace this query criteria:
```
created_at + delay <= now()
```
with
```
scheduled_at  <= now()
```

**Reply by maxcountryman** [reply]
"Do it at" is a function of the scheduler.

"Do it after" doesn't require the scheduler.

So to answer your question, the scheduler is needed whenever the first, "do it at", is required.

That said, I think the point you're making is that we could use a timestamp instead for the "do it after" semantic.

I don't disagree but I would prefer to not do time zone conversion outside of Postgres and so the decision to operate over intervals within the app was intentional.

If you think there's a strong reason to not do this I'm happy to think more about that. 🙂 As I currently understand, it would remove the arithmetic of `now() + delay` but I'm not sure that's a significant win and it would come at the cost of serializing timestamps between the application layer and Postgres which I'd like to try to avoid if it's not necessary.

**Reply by kirillsalykin** [reply]
tbh, there is no strong reason, it just feels more natural.
but probably you are right, it is not worth the trouble, `scheduled_at` can be implemented in terms of `created_at` + `delay`

**Reply by maxcountryman** [reply]
> it is not worth the trouble

I'm particularly concerned about time zones here because Postgres is already going to be converting from stored UTC to whatever the cluster is configured with. Rather than try to account for that in the application, I'd prefer to just let Postgres handle all conversions to the extent possible.

**Reply by kirillsalykin** [reply]
AFAIK, psql stores everything in UTC.

**Reply by kirillsalykin** [reply]
anyway, i dont think it worth it. lets use delay.

**Reply by maxcountryman** [reply]
Yes it does, but once read it's converted to the locale the cluster is to set (that could be UTC or anything else).
